pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/DistributedTaskLockManager.java

```diff
@@ -0,0 +1,288 @@
+        _controllerInstanceId);
+
+    try {
+      // Check if task generation is already in progress
+      if (isTaskGenerationInProgress(tableNameWithType)) {
+        LOGGER.info("Task generation already in progress for: {} by this or another controller", tableNameWithType);
+        return null;
+      }
+
+      // Try to acquire the lock using ephemeral node
+      TaskLock lock = tryAcquireSessionBasedLock(tableNameWithType);
+      if (lock != null) {
+        LOGGER.info("Successfully acquired task generation lock for table: {} by controller: {}", tableNameWithType,
+            _controllerInstanceId);
+        return lock;
+      } else {
+        LOGGER.warn("Could not acquire lock for table: {} - another controller must hold it", tableNameWithType);
+      }
+    } catch (Exception e) {
+      LOGGER.error("Error while trying to acquire task lock for table: {}", tableNameWithType, e);
+    }
+    return null;
+  }
+
+  private String getLockPath(String tableNameForPath) {
+    return ZKMetadataProvider.constructPropertyStorePathForMinionTaskGenerationLock(tableNameForPath);
+  }
+
+  /**
+   * Releases a previously acquired session-based lock and marks task generation as completed.
+   *
+   * @param lock the lock to release
+   * @return true if successfully released, false otherwise
+   */
+  public boolean releaseLock(TaskLock lock) {
+    if (lock == null) {
+      return true;
+    }
+
+    String tableNameWithType = lock.getTableNameWithType();
+    String lockNode = lock.getLockZNodePath();
+
+    // Remove the ephemeral lock node
+    boolean status = true;
+    if (lockNode != null) {
+      try {
+        if (_propertyStore.exists(lockNode, AccessOption.EPHEMERAL)) {
+          status = _propertyStore.remove(lockNode, AccessOption.EPHEMERAL);
+          LOGGER.info("Tried to removed ephemeral lock node: {}, for table: {} by controller: {}, removal success: {}",
+              lockNode, tableNameWithType, _controllerInstanceId, status);
+        } else {
+          LOGGER.warn("Ephemeral lock node: {} does not exist for table: {}, nothing to remove",
+              lockNode, tableNameWithType);
+        }
+      } catch (Exception e) {
+        status = false;
+        LOGGER.warn("Exception while trying to remove ephemeral lock node: {}", lockNode, e);
+      }
+    } else {
+      LOGGER.warn("Lock node path seems to be null for task lock: {}, treating release as a no-op", lock);
+    }
+
+    return status;
+  }
+
+  /**
+   * Force release the lock without checking if any tasks are in progress
+   */
+  public boolean forceReleaseLock(String tableNameWithType) {
+    LOGGER.info("Trying to force release the lock for table: {}", tableNameWithType);
+    String lockPath = getLockPath(tableNameWithType);
+
+    boolean released = true;
+    if (!_propertyStore.exists(lockPath, AccessOption.EPHEMERAL)) {
+      LOGGER.info("No lock ZNode: {} found for table: {}, nothing to force release", lockPath, tableNameWithType);
```

Comment:
we can probably throw an exception. that way client will know what actually happened. otherwise returning either true / false in such a case might be a false signal. 

similarly an exception can be thrown if _propertyStore.remove returns false

Reply:
done